[18.0][MIG] sale_order_invoice_amount: Migration to 18.0

sbiosca-s73 opened the pull request:
## What this module does in 18.0

This module adds the **amount_invoiced** and **amount_to_invoice** fields to sale order views:

- In the **form view**, both fields are displayed in the tax totals summary, above the total order amount.
- In the **list view**, both fields are available as optional user fields.

## Version history note

In version 17 and earlier, this module introduced **amount_invoiced** and **amount_to_invoice** as new computed fields on sale orders.  
Since version 18, these fields are already provided by Odoo core. The module now focuses on **UI visibility only**, exposing those existing fields in sale order views.

MateuGForgeFlow committed:
[ADD] sale_order_invoice_amount

MateuGForgeFlow committed:
[ADD] sale_order_invoice_amount

JasminSForgeFlow committed:
[IMP] sale_order_invoice_amount: black, isort, prettier

JasminSForgeFlow committed:
[IMP] sale_order_invoice_amount: black, isort, prettier

JasminSForgeFlow committed:
[MIG] sale_order_invoice_amount: Migration to 14.0

JasminSForgeFlow committed:
[MIG] sale_order_invoice_amount: Migration to 14.0

JasminSForgeFlow committed:
[MIG] sale_order_invoice_amount: Migration to 15.0

JasminSForgeFlow committed:
[MIG] sale_order_invoice_amount: Migration to 15.0

cesar-tecnativa committed:
[MIG] sale_order_invoicing_finished_task: Migration to 15.0

cesar-tecnativa committed:
[MIG] sale_order_invoicing_finished_task: Migration to 15.0

Commit:
[MIG] sale_order_invoice_amount: Migration to 16.0

Commit:
[MIG] sale_order_invoice_amount: Migration to 16.0

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_invoice_amount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_invoice_amount/

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_invoice_amount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_invoice_amount/

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

Commit:
Added translation using Weblate (French)

Commit:
Added translation using Weblate (French)

Commit:
Translated using Weblate (French)

Currently translated at 100.0% (5 of 5 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_invoice_amount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_invoice_amount/fr/

Commit:
Translated using Weblate (French)

Currently translated at 100.0% (5 of 5 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_invoice_amount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_invoice_amount/fr/

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (5 of 5 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_invoice_amount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_invoice_amount/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (5 of 5 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_invoice_amount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_invoice_amount/it/

DavidJForgeFlow committed:
[IMP] sale_order_invoice_amount: make sure to respect currencies

DavidJForgeFlow committed:
[IMP] sale_order_invoice_amount: make sure to respect currencies

DavidJForgeFlow committed:
[IMP] sale_order_invoice_amount: uninvoiced amount checks qty invoiced not total amount

DavidJForgeFlow committed:
[IMP] sale_order_invoice_amount: uninvoiced amount checks qty invoiced not total amount

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
[BOT] post-merge updates

MeritxellAForgeFlow committed:
[IMP] sale_order_invoice_amount: pre-commit stuff

MeritxellAForgeFlow committed:
[IMP] sale_order_invoice_amount: pre-commit stuff

MeritxellAForgeFlow committed:
[MIG] sale_order_invoice_amount: Migration to 17.0

MeritxellAForgeFlow committed:
[MIG] sale_order_invoice_amount: Migration to 17.0

rousseldenis commented:
/ocabot migration sale_order_invoice_amount

Reyes4711-S73 commented:
LGTM

BhaveshHeliconia reviewed (approved):
Functional review LGTM!

sbiosca-s73 commented:
@rousseldenis can you merge please?

rousseldenis commented:
@sbiosca-s73 No need of migration script for fields renaming ?

Commit:
[UPD] Update sale_order_invoice_amount.pot

Commit:
[UPD] Update sale_order_invoice_amount.pot

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
[BOT] post-merge updates

DavidJForgeFlow commented:
Hi @sbiosca-s73, there were some changes in v17.0 removing amount_invoiced and amount_to_invoice. Please could you get the recent changes to have both versions aligned?

17.0 PR: https://github.com/OCA/sale-workflow/pull/3537

DavidJForgeFlow commented:
Superseded in #3955 as it needs more changes.

rousseldenis commented:
@sbiosca-s73 Do you agree we move to https://github.com/OCA/sale-workflow/pull/3955 ?

sbiosca-s73 commented:
@DavidJForgeFlow  Sorry, I missed your suggestion.

@rousseldenis I have seen the new PR, and I agree those canges look like correct migration, but i have a doubt about _compute_amount_to_invoice and _compute_amount_invoiced. @DavidJForgeFlow according with your comment, you removed this computes because Odoo base already handles the functionality. David, I see URL change Odoo base in your PR, but i'm not clear on how it replaces the functionality from OCA module. Could you please explain me why the OCA logic is no longer necessary? Once we clarify this, we can decide whether to keep the changes in your PR or apply them to this.

Thanks!

DavidJForgeFlow commented:
> @DavidJForgeFlow Sorry, I missed your suggestion.
> 
> @rousseldenis I have seen the new PR, and I agree those canges look like correct migration, but i have a doubt about _compute_amount_to_invoice and _compute_amount_invoiced. @DavidJForgeFlow according with your comment, you removed this computes because Odoo base already handles the functionality. David, I see URL change Odoo base in your PR, but i'm not clear on how it replaces the functionality from OCA module. Could you please explain me why the OCA logic is no longer necessary? Once we clarify this, we can decide whether to keep the changes in your PR or apply them to this.
> 
> Thanks!

Hi @sbiosca-s73,
The computation of both fields was done in the migration to v17.0 (that's why we deleted the invocied_amount and uninvoiced_amount). However the amount_to_invoice was still being computed taking into account the amount left and ignoring the quantity of products invoiced. For example, if I sell 2 units at 15 €/unit, If I has an invoice of 1 unit with amount 20€, the amount_to_invoice was 10€. In this module the amount was 15€ as one unit was left to invoice. 
This was changed in v18.0 in the commit I commented. Now the computation is the same and therefore we can remove the method (also taking into account the changes in currencies are also treated).

sbiosca-s73 commented:
@DavidJForgeFlow @rousseldenis I have implemented the correct migration changes that David mentioned, so we can close PR #3955 and we can respect this PR as the initial migration. Thank you.

DavidJForgeFlow commented on sale_order_invoice_amount/static/src/xml/tax_totals.xml:
```diff
@@ -0,0 +1,37 @@
+                <td class="o_td_label">
+                    <label
+                        class="o_form_label o_tax_total_label"
+                    >Uninvoiced Amount</label>
```
To keep the Odoo standard naming, maybe it's better to adapt this string to Amount to Invoice

DavidJForgeFlow commented on sale_order_invoice_amount/static/src/xml/tax_totals.xml:
```diff
@@ -0,0 +1,37 @@
+                <td class="o_td_label">
+                    <label
+                        class="o_form_label o_tax_total_label"
+                    >Invoiced Amount</label>
```
Maybe it's better to keep the standard naming: Amount Invoiced

DavidJForgeFlow commented on sale_order_invoice_amount/views/sale_order_view.xml:
```diff
@@ -0,0 +1,33 @@
+                <field name="amount_invoiced" sum="Invoiced Amount" optional="hide" />
+                <field
+                    name="amount_to_invoice"
+                    sum="Uninvoiced Amount"
```
Same as above

DavidJForgeFlow reviewed (commented):
Great! Thanks @sbiosca-s73, I will then close the other PR. COuld you please review this?

sbiosca-s73 commented:
@DavidJForgeFlow done! Thank you

DavidJForgeFlow commented on sale_order_invoice_amount/views/sale_order_view.xml:
```diff
@@ -0,0 +1,33 @@
+    <record id="view_quotation_list_with_onboarding_invoiced_amount" model="ir.ui.view">
+        <field name="name">quotation.list.with.onboarding.invoiced_amount</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_quotation_tree_with_onboarding" />
```
You should inherit sale.view_quotation_tree

sbiosca-s73 commented:
@DavidJForgeFlow done, thank you!

DavidJForgeFlow committed:
[IMP] sale_order_invoice_amount: get correct api.depends

There are some cases when the amount to invoice is not actually updated when it needs to, so this adds the dependencies that base method has and adding the state of the SO to be able to get it updated when SO is confirmed.

DavidJForgeFlow committed:
[IMP] sale_order_invoice_amount: get correct api.depends

There are some cases when the amount to invoice is not actually updated when it needs to, so this adds the dependencies that base method has and adding the state of the SO to be able to get it updated when SO is confirmed.

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
[BOT] post-merge updates

Reyes4711-S73 commented:
@OCA/crm-sales-marketing-maintainers Please, can you review/merge this PR?

LoisRForgeFlow reviewed (changes requested):
This migration is clearly incorrect as it is using commits that does not correspond with the ones in v17.

You can have a looks a the commit "[MIG] sale_order_invoice_amount: Migration to 17.0" in this branch and in the merged version in 17.0 at OCA:

https://github.com/OCA/sale-workflow/pull/3819/changes/35a26e056fa03a6fa610931c7bf2abed0cc4e753

https://github.com/OCA/sale-workflow/commit/a00d04bdc7aca90aa99b4047c3778b81b839db3b#diff-9531a658599091a48247513901354d68c1dad55fd24681abf99b9cca39922f00R10

They are not the same...

And also this misses commits from 17.0, like this one https://github.com/OCA/sale-workflow/commit/25a3ea07805199fb63dbd5f57711b21f6313349e

sbiosca-s73 committed:
[IMP] sale_order_invoice_amount: pre-commit auto fixes

sbiosca-s73 committed:
[IMP] sale_order_invoice_amount: pre-commit auto fixes

sbiosca-s73 committed:
[MIG] sale_order_invoice_amount: Migration to 18.0

sbiosca-s73 committed:
[MIG] sale_order_invoice_amount: Migration to 18.0

mmequignon commented:
> This migration is clearly incorrect as it is using commits that does not correspond with the ones in v17.
> 
> You can have a looks a the commit "[MIG] sale_order_invoice_amount: Migration to 17.0" in this branch and in the merged version in 17.0 at OCA:
> 
> [35a26e0](https://github.com/OCA/sale-workflow/commit/35a26e056fa03a6fa610931c7bf2abed0cc4e753)
> 
> [a00d04b#diff-9531a658599091a48247513901354d68c1dad55fd24681abf99b9cca39922f00R10](https://github.com/OCA/sale-workflow/commit/a00d04bdc7aca90aa99b4047c3778b81b839db3b#diff-9531a658599091a48247513901354d68c1dad55fd24681abf99b9cca39922f00R10)
> 
> They are not the same...
> 
> And also this misses commits from 17.0, like this one [25a3ea0](https://github.com/OCA/sale-workflow/commit/25a3ea07805199fb63dbd5f57711b21f6313349e)

I agree.
Please @sbiosca-s73, stick to this https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-18.0
Thank you!

sbiosca-s73 commented:
@LoisRForgeFlow It is expected that some v17 commits were missing from this migration PR, as I originally performed the migration and opened the PR on July 17, 2025. The changes you mentioned were made subsequently, so I had not updated the migration PR until now. I have already applied those changes.

@mmequignon The problem is due to the timing of the changes, it is not an issue with the migration process.

Changes done!
Thank you.

LoisRForgeFlow commented:
@sbiosca-s73 Thanks for the clarification and quick fixing!

Now the migration commit looks much different and less straight forward as it is removing options that were added in v17, I would request @Reyes4711-S73 @DavidJForgeFlow and @BhaveshHeliconia to check again and update their review.

BhaveshHeliconia reviewed (approved):
Functional review LGTM!

alexey-pelykh commented on sale_order_invoice_amount/models/sale_order.py:
```diff
@@ -0,0 +1,39 @@
+                    ),
+                }
+            )
+        return res
```
This file only overrides `_compute_tax_totals()` to add formatted amounts to the widget, but is **missing** the core overrides from the 17.0 version:

- `_compute_amount_invoiced()` -- contains the quantity-based invoiced amount calculation with multi-currency support and the `enable_amount_invoiced_based_on_quantity` company setting check
- `_compute_amount_to_invoice()` -- contains the quantity-based uninvoiced amount calculation with the company setting check

Without these, the module does not change any computation logic -- it merely re-displays the standard Odoo `amount_invoiced` and `amount_to_invoice` values in the tax totals section, which Odoo may already do natively.

See: https://github.com/OCA/sale-workflow/blob/17.0/sale_order_invoice_amount/models/sale_order.py

alexey-pelykh commented on sale_order_invoice_amount/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    "category": "Sales",
+    "license": "LGPL-3",
+    "summary": "Display the invoiced and uninvoiced total in the sale order",
+    "depends": ["sale"],
```
The summary is from the 16.0 version. The 17.0 version updated this to:
```python
"summary": "Improves Sales Order invoiced/to invoice amount calculation "
"based on product quantities when the company setting is enabled.",
```
Also, `data` should include `views/sale_order_config_settings.xml` (once the config settings feature is ported from 17.0).

alexey-pelykh commented on sale_order_invoice_amount/tests/test_sale_order_invoice_amount.py:
```diff
@@ -0,0 +1,429 @@
+        # Products
+        product_model = cls.env["product.product"]
+        cls.product_1 = product_model.create(
+            {"name": "Desk Combination", "type": "consu"}
```
Missing the company setting enablement from the 17.0 tests:
```python
cls.env.company.write(
    {
        "enable_amount_invoiced_based_on_quantity": True,
    }
)
```
Without this (and without the company setting feature being ported), these tests are exercising standard Odoo computation, not the module's custom logic.

alexey-pelykh reviewed (changes requested):
## Review: REQUEST_CHANGES

This migration has critical issues, aligning with the concerns already raised by @LoisRForgeFlow. The PR appears to migrate from the **16.0** version directly to 18.0, skipping the substantial rework done in **17.0**.

### Critical Issues

**1. Missing core business logic from 17.0**

The 17.0 version overrides `_compute_amount_invoiced()` and `_compute_amount_to_invoice()` (the native Odoo 17+ fields), which contain the actual quantity-based calculation logic that is the whole point of this module. The current PR only overrides `_compute_tax_totals()` to add formatting to the tax totals widget, but does NOT include the custom computation logic. Without these overrides, the module only displays the standard Odoo-computed amounts in the tax totals section -- it does not actually change HOW the amounts are calculated.

**2. Missing company configuration feature (from 17.0)**

The 17.0 version introduced:
- `res_company.py` with `enable_amount_invoiced_based_on_quantity` Boolean field
- `sale_order_config_settings.py` with the related field on `res.config.settings`
- `views/sale_order_config_settings.xml` for the settings UI
- `data` in `__manifest__.py` includes `views/sale_order_config_settings.xml`

This allows the quantity-based calculation to be toggled per company. The current PR has none of this.

**3. `account` dependency removed without justification**

The 16.0 manifest depends on `["sale", "account"]`. The 17.0 manifest depends on `["sale"]` (correctly, since the `account` dependency is pulled transitively). The current PR also only depends on `["sale"]`, which is fine, but this should be intentional.

**4. Tests do not enable the company setting**

The 17.0 tests include `cls.env.company.write({"enable_amount_invoiced_based_on_quantity": True})` in `setUpClass`. The current tests do not, because the feature does not exist in this PR. This means the tests are testing standard Odoo behavior, not the module's custom logic.

**5. Summary in `__manifest__.py` is stale**

The 17.0 manifest updated the summary to: *"Improves Sales Order invoiced/to invoice amount calculation based on product quantities when the company setting is enabled."* The current PR still uses the old 16.0 summary.

### Recommendation

Please rebase this migration on the **17.0 branch** as the starting point (not 16.0). The 17.0 version is the correct baseline for an 18.0 migration. Key files to port from 17.0:
- `models/res_company.py`
- `models/sale_order_config_settings.py`
- `models/sale_order.py` (with `_compute_amount_invoiced` and `_compute_amount_to_invoice` overrides)
- `views/sale_order_config_settings.xml`
- Updated `__manifest__.py` (data list, summary)
- Updated tests (with company setting enabled)

sbiosca-s73 commented:
@alexey-pelykh Regarding your points about using v17 as the baseline instead of v16, the specific features you mentioned—such as `_compute_amount_invoiced`, `enable_amount_invoiced_based_on_quantity`, etc.—were modified based on feedback from @DavidJForgeFlow. 
If you review the earlier conversation in this thread, you will see that David suggested modifying this PR and even opened an alternative one in case we decided to move the changes there. The current implementation was intentionally left this way because we agreed with the approach and logic David proposed in his PR (https://github.com/OCA/sale-workflow/pull/3955). Could you please review the previous discussions in the thread? 
Given that context, I do not believe the changes you are suggesting from v17 are necessary for the v18 migration. 
Thank you very much!

alexey-pelykh reviewed (commented):
@sbiosca-s73 Thank you for pointing me to the earlier discussion with @DavidJForgeFlow and @LoisRForgeFlow. I should have reviewed the full thread history before raising those points.

I see that the company setting and custom compute overrides from 17.0 were intentionally scoped out based on the agreed approach. That context makes the current implementation make sense.

Since #3955 is closed: are the quantity-based calculation features planned for a separate follow-up PR on 18.0, or intentionally dropped? Just want to make sure the module's roadmap is clear for future contributors.

Tisho99 commented on sale_order_invoice_amount/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,2 @@
+This module adapts the computation of invoiced and to invoice amounts in Sales Orders.
+When the company option "Enable amount invoiced based on quantity" is activated, the values are calculated from the quantities invoiced vs ordered, instead of relying only on invoice monetary amounts.
```
I think that the readme should be refactored.

The second line appears to need to be deleted. That option has been removed from this module, and i haven't found it in the Odoo core modules

It has also been a lot of changes and discussion about this module, and the readme should clarify it in a little summary

Tisho99 replied:
I'll put this:

This module adds the amount_invoiced and amount_to_invoice fields to the sale order views:

In the form view, those fields are added in the tax totals summary, above the total order amount
In the list view, the fields are added as user optional fields.

Tisho99 replied:
I'll also add this to the roadmap or the changelog, and the PR's description:

In version 17 of this module and earlier ones, the amount_invoiced and amount_to_invoice fields were added to the sale order model as new computed fields. Since version 18, these fields have been added directly by Odoo, but they remain hidden in the interface, so this module now just displays the fields.

Tisho99 reviewed (approved):
Hello.

I performed a thorough review of the module and the entire thread.

The migration is correct, so I approve the pull request. All the line removals are correct, as the functionality is now covered by Odoo. The module now has a more purely visual approach than in previous versions, but it remains very useful for viewing invoiced amounts.

However, I see deficiencies in the README and the pull request documentation. This migration has evolved significantly over time, and the changes made haven't been properly documented here. This lack of documentation could also confuse new reviewers of the thread.

I believe that by addressing this, the pull request would be fully ready for merging, as it has many approvals, and the comments from alexey-pelykh would be resolved with the latest clarifications made in the thread.

@sbiosca-s73, could you please implement the changes I'm requesting or provide your feedback?

If for whatever reason you can't do it, I'd like to try to finish this work in a separate OCA PR.

Thank you for your work!!

sbiosca-s73 commented:
@Tisho99  Thank you for the review and for your feedback regarding these points. I have updated the module documentation to align with the finalized behavior for version 18.0 by removing the outdated description of custom computation logic and clarifying that the module now exclusively exposes Odoo core fields within the user interface. These updates ensure that the README and pull request documentation accurately reflect the actual scope of this migration.

Tisho99 commented:
Hello @sbiosca-s73 

Thank you for applying the suggested changes, now it looks even better.

I'll try to ask if a PSC can review this

ValentinVinagre reviewed (approved):
lgtm 👍🏻

HaraldPanten reviewed (approved):
/ocabot merge nobump
